### PR TITLE
Emit completion events

### PR DESCRIPTION
**Pull request by bryanlandia**
Emit started, completed and revoked events using eventtracking's Tracker.  Emits a "generic" tracking event as well as another destined for BI (Segment) integration (if Segment enabled).

Emit `edx.bi.completion.user.{aggregator_type}.completed` (aggregator type is like 'course', 'chapter', 'vertical', 'sequential') when an `Aggregator` is *created* or *updated* (using the `AggregationUpdater's` `update()` method only)  and the percent complete changes to 100 — i.e., emit only once per completion.  Emit `edx.completion.aggregator.completed` with the same criteria.

Emit `edx.bi.completion.user.{aggregator_type}.started`when an `Aggregator` is *created* (using the `AggregationUpdater's` `update()` method only)  and the percent is above 0% .  Emit `edx.completion.aggregator.started` with the same criteria.

Emit `edx.bi.completion.user.{aggregator_type}.revoked`when an `Aggregator` is *updatde* (using the `AggregationUpdater's` `update()` method only)  and the percent changes from 100% to something less .  Emit `edx.completion.aggregator.revoked` with the same criteria.

It's configurable with defaults set and overridable via `SiteConfiguration`

`COMPLETION_AGGREGATOR_ENABLE_TRACKING` to enable. Defaults to False.  
Reviewer:  Should we use a Waffle swtich instead?  That's a little weirder to deal with for Site-specific config.

`COMPLETION_AGGREGATOR_TRACKED_BLOCK_TYPES` to only track certain block types.  Defaults to `['course', 'chapter']` since lower-level Aggregators are likely to be of less interest.

### CODE REVIEW

**Comment by bryanlandia** [comment]
I've left this in Draft since I haven't added any tests.  I will try to get some written but would be good to get some eyes on this in the meantime.

**Comment by bryanlandia** [comment]
I've left this in Draft since I haven't added any tests.  I will try to get some written but would be good to get some eyes on this in the meantime.

**Comment by bryanlandia** [comment]
Some tests in place now, all existing tests passing, ready for review.

File: completion_aggregator/settings/common.py [hunk 1]
```diff
@@ -16,3 +16,5 @@ def plugin_settings(settings):
         'vertical',
     }
     settings.COMPLETION_AGGREGATOR_ASYNC_AGGREGATION = False
+    settings.COMPLETION_AGGREGATOR_ENABLE_TRACKING = False
+    settings.COMPLETION_AGGREGATOR_TRACKED_BLOCK_TYPES = ['course', 'chapter', ]
```
Comment (OmarIthawi) [inline completion_aggregator/settings/common.py]:
```suggestion
    settings.COMPLETION_AGGREGATOR_TRACKED_BLOCK_TYPES = {'course', 'chapter'}
```

File: completion_aggregator/tracking.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+"""
+
+from eventtracking import tracker
+
+from . import compat
+
+
+TRACKER_BI_EVENT_NAME_FORMAT = u'edx.bi.user.{agg_type}.{event_type}'
+TRACKER_EVENT_NAME_FORMAT = u'edx.completion.aggregator.{event_type}'
+TRACKER_VALID_EVENT_TYPES = ('completed', 'started', )
```
Comment (OmarIthawi) [inline completion_aggregator/tracking.py]: Set is better, although there might be no performance difference.

```suggestion
TRACKER_VALID_EVENT_TYPES = {'completed', 'started'}
```

File: completion_aggregator/tracking.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    agg_type = aggregator.aggregation_name
+    block_id = str(aggregator.block_key)
+    course_id = str(aggregator.course_key)
+    percent = aggregator.percent * 100
+    earned = aggregator.earned
+    possible = aggregator.possible
+
+    # BI event if we have a SEGMENT integration
+    if compat.get_segment_key() is not None:
```
Comment (OmarIthawi) [inline completion_aggregator/tracking.py]:
```suggestion
    if compat.get_segment_key():
```

File: completion_aggregator/tracking.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
```
Comment (OmarIthawi) [inline completion_aggregator/tracking.py]:
```suggestion
```

File: completion_aggregator/tracking.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        'block_type': agg_type,
+    })
+
+
+def track_aggregation_events(aggregator, is_new=False):
+    """
+    If event tracking feature is enabled and is a trackable type, call function to emit tracking events.
+    Keep in mind that the aggregator may not have been saved to the database yet.
+    """
+    if compat.is_tracking_enabled() is not True:
```
Comment (OmarIthawi) [inline completion_aggregator/tracking.py]:
```suggestion
    if not compat.is_tracking_enabled():
```

File: test_utils/test_app/models.py [hunk 6]
```diff
@@ -3,13 +3,18 @@
...
+from collections import OrderedDict
+
 from opaque_keys.edx.django.models import CourseKeyField
 
 from django.conf import settings
 from django.contrib.auth.models import User
 from django.db import models
 
 
+TRACKING_TESTING_COURSE_IDS = ('course-v1:Appsembler+AggEvents101+2020', )
```
Comment (OmarIthawi) [inline test_utils/test_app/models.py]:
```suggestion
TRACKING_TESTING_COURSE_IDS = {'course-v1:Appsembler+AggEvents101+2020'}
```

File: tests/test_core.py [hunk 7]
```diff
@@ -52,41 +52,46 @@ def setUp(self):
...
-            self.addCleanup(patch.stop)
-        user = get_user_model().objects.create(username='saskia')
-        self.course_key = CourseKey.from_string('course-v1:edx+course+test')
+        for compat_import in (
+                'completion_aggregator.core.compat',
+                'completion_aggregator.tracking.compat',
+        ):
+            patcher = mock.patch(compat_import, stubcompat)
+            patcher.start()
+            # self.addCleanup(patcher.__exit__, None, None, None)  # TODO: second patcher won't work if uncommented. why?
```
Comment (OmarIthawi) [inline tests/test_core.py:74]:
```suggestion
```

**Review by OmarIthawi — COMMENTED** [review]
I've finished a first pass, I'll take another look soon.

File: tests/test_core.py (hunk 7 above)
Comment (bryanlandia) [reply tests/test_core.py:74]: I want to keep this note for now

**Comment by bryanlandia** [comment]
Thanks for the review @OmarIthawi I committed most of the suggestions.

File: completion_aggregator/tracking.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    agg_type = aggregator.aggregation_name
+    block_id = str(aggregator.block_key)
+    course_id = str(aggregator.course_key)
+    percent = aggregator.percent * 100
+    earned = aggregator.earned
+    possible = aggregator.possible
+
+    # BI event if we have a SEGMENT integration
+    if compat.get_segment_key():
```
Comment (OmarIthawi) [inline completion_aggregator/tracking.py:45]: Question for @melvinsoft. Does this behaviour match the edX Platform check for segment?

File: completion_aggregator/tracking.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            try:
+                assert course_block_struct['block_type'] == 'course'
+                course_name = course_block_struct['display_name']
+            except AssertionError:  # this shouldn't happen
+                course_name = course_id
+        except (compat.coursestructure_model().DoesNotExist, KeyError):
+            course_name = course_id
+            block_name = block_id
+
+        tracker.emit(bi_event_name, {
```
Comment (OmarIthawi) [inline completion_aggregator/tracking.py:65]: Question for @melvinsoft. Does this post the event for both the customer and platform-wide Segment?

File: completion_aggregator/tracking.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            'label': '{} {} {}'.format(agg_type, block_name, event_type_label),
+            'course_id': course_id,
+            'block_id': block_id,
+            'completion_percent': percent,
+            'course_name': course_name,
+            'block_name': block_name,
+        })
+
+    # generic tracking event
+    event_name = TRACKER_EVENT_NAME_FORMAT.format(event_type=event_type)
```
Comment (OmarIthawi) [inline completion_aggregator/tracking.py:75]: @bryanlandia what's the generic tracking event and how different is it from `tracker.emit(bi_event_name)` above?

File: tests/test_core.py [hunk 11]
```diff
@@ -52,41 +52,46 @@ def setUp(self):
         """
         super(AggregationUpdaterTestCase, self).setUp()
         self.agg_modified = now() - timedelta(days=1)
-        course_key = CourseKey.from_string('course-v1:edx+course+test')
+        self.course_key = CourseKey.from_string('course-v1:edx+course+test')
```
Comment (OmarIthawi) [inline tests/test_core.py]: Keep the diff minimal

```suggestion
        course_key = CourseKey.from_string('course-v1:edx+course+test')
        self.course_key = course_key
```

File: tests/test_core.py [hunk 12]
```diff
@@ -52,41 +52,46 @@ def setUp(self):
...
+            self.course_key.make_usage_key('chapter', 'course-chapter'),
+            self.course_key.make_usage_key('html', 'course-chapter-html'),
+            self.course_key.make_usage_key('html', 'course-html0'),
+            self.course_key.make_usage_key('html', 'course-html1'),
+            self.course_key.make_usage_key('html', 'course-html2'),
+            self.course_key.make_usage_key('html', 'course-html3'),
+            self.course_key.make_usage_key('other', 'course-other'),
+            self.course_key.make_usage_key('hidden', 'course-hidden0'),
+            self.course_key.make_usage_key('html', 'course-other-html4'),
+            self.course_key.make_usage_key('hidden', 'course-other-hidden1'),
```
Comment (OmarIthawi) [inline tests/test_core.py]: Reduce future merge issues.

```suggestion
            course_key.make_usage_key('course', 'course'),
            course_key.make_usage_key('html', 'course-html0'),
            course_key.make_usage_key('html', 'course-html1'),
            course_key.make_usage_key('html', 'course-html2'),
            course_key.make_usage_key('html', 'course-html3'),
            course_key.make_usage_key('other', 'course-other'),
            course_key.make_usage_key('hidden', 'course-hidden0'),
            course_key.make_usage_key('html', 'course-other-html4'),
            course_key.make_usage_key('hidden', 'course-other-hidden1'),
```

**Review by OmarIthawi — APPROVED** [review]
Thanks @bryanlandia. I did another review and I think it's good to go. Thank you for adding tests.

I think we need the following before we merge this PR:

 - [ ] @melvinsoft to verify that this works well without our [Tahoe custom Segment BI feature](https://github.com/appsembler/edx-platform/pull/284)

Additionally there are few points that would be good to have:

 - Could you please explain to me what's the difference between a BI event and a normal tracking event?
 - Could you please enable TravisCI?

**Comment by amirtds** [comment]
@bryanlandia I did review the PR 2 times today and couldn't find something wrong besides some minor PEP8 fixes.

**Review by amirtds — APPROVED** [review]

File: completion_aggregator/tracking.py (hunk 9 above)
Comment (bryanlandia) [reply completion_aggregator/tracking.py:65]: I am seeing two copies of the event in my test Segment account when I have the LMS setting and SiteConfiguration Segment key set to the same write key.

File: completion_aggregator/tracking.py (hunk 10 above)
Comment (bryanlandia) [reply completion_aggregator/tracking.py:75]: @OmarIthawi "Generic events" (my term to distinguish from BI) are ones that show up in the tracking logs via `LoggerBackend` or can get sent to any number of other defined backends.  They contain a lot of detail and are structured in a way that has been stable for at least several edx-platform releases.  

"BI events" follow a model used in edx-platform for the `edx.bi.` namespace events.  These are structured differently to work with Segment.  To be usable in Segment Destinations (Marketo, Salesforce, Intercom, etc.), event properties need to be at the top level of the JSON structure and shouldn't re-use reserved names from Segment like `context`.  You'll see edX has taken a poor approach to getting around the unusable event structure at present produced by event-tracking and/or lms.djangoapps.track by just calling Segment's Python API methods directly `analytics.track(), analytics.identify()` etc. 👎 👎  All those really should be deleted from edx-platform.  

How it should work is that the `SegmentBackend` already defined as an event backend in event-tracking should actually format  the events in a way that's usable in Segment.  That's what I've done with [this PR to event-tracking](https://github.com/appsembler/event-tracking/pull/1) (while maintaining backward compatibility).

**Comment by bryanlandia** [comment]
Checking the Travis failures.  Some of the tox setup had issues in the original branch, too.  Once approved, I'll rebase and squash before merging.

**Comment by bryanlandia** [comment]
I looked into the travis failures. There are issues in the tox.ini and test env requirements caused by Django version.  I'm hesitant to make a ton of changes in the test env requirements files, at least for this PR, since it produces a large diff.  Basically, if you comment out the Django version spec the testenv tests run and pass.

File: completion_aggregator/tracking.py (hunk 9 above)
Comment (melvinsoft) [reply completion_aggregator/tracking.py:65]: Yes, it should emit the event in both places.

**Comment by melvinsoft** [comment]
@bryanlandia I replied a couple of question, I don't the capacity to fully review the PR since I'm not familiar with completion yet, but I've replied the specific questions that Omar posted and posted a new one for you.

**Comment by bryanlandia** [comment]
Thanks @melvinsoft I think you might not have saved something since I didn't see any new question from you

File: completion_aggregator/tracking.py (hunk 10 above)
Comment (bryanlandia) [reply completion_aggregator/tracking.py:75]: that PR to event-tracking has been replaced with one to do the same for now in a new module in our appsembler Django app. https://github.com/appsembler/edx-platform/pull/590

**Comment by bryanlandia** [comment]
Argh, I was supposed to rebase and squash before merging.  I'm going to do that but directly on appsembler/hawthorn/master now

**Pull request merged**